opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -2290,6 +2290,73 @@ namespace detail {
 
 
 
+
+    template <class Grid, class WellModel, class Implementation>
+    std::vector<V>
+    BlackoilModelBase<Grid, WellModel, Implementation>::
+    computeFluidInPlace(const ReservoirState& x,
+                        const WellState& xw,
+                        const std::vector<int>& fipnum)
+    {
+        using namespace Opm::AutoDiffGrid;
+        const int nc = numCells(grid_);
+        const int np = x.numPhases();
+
+        SolutionState state(np);
```

Comment:
This part recomputes lots of data already computed. Instead, we should refactor the model class a little bit to store the necessary quantities (some are already stored in the rq_ member).